src/vsr/grid.zig

```diff
@@ -765,76 +865,57 @@ pub fn GridType(comptime Storage: type) type {
             // Remove the "root" read so that the address is no longer actively reading / locked.
             grid.read_queue.remove(read);
 
-            // A valid block filled by storage means the reads for the address can be resolved.
-            if (read_block_valid(cache_block.*, .{
+            const result = read_block_validate(block.*, .{
                 .address = read.address,
                 .checksum = read.checksum,
-                .block_type = read.block_type,
-            })) {
-                grid.read_block_resolve(read, cache_block.*);
+            });
+
+            if (result == .valid) {
+                grid.read_block_resolve(read, .{ .valid = block.* });
                 return;
             }
 
+            const header = mem.bytesAsValue(vsr.Header, block.*[0..@sizeOf(vsr.Header)]);
+            log.err(
+                "{s}: expected address={} checksum={}, found address={} checksum={}",
+                .{
+                    @tagName(result),
+                    read.address,
+                    read.checksum,
+                    header.op,
+                    header.checksum,
+                },
+            );
+
             // Don't cache a corrupt or incorrect block.
```

Comment:
We cache the block conditionally on `read.cache_wire`, but here we remove it unconditionally. This feels wrong: if a scrubber finds a bad block, but it is actually still cached, we don't want to evict the cached copy

Reply:
Great catch!